File: sources/droid/AdMobWapperApp/admob-wrapper/src/main/java/it/amdev/admob/wrapper/AdMobManager.kt
package it.amdev.admob.wrapper

import android.app.Activity
import android.content.Context
import com.google.android.libraries.ads.mobile.sdk.MobileAds
import com.google.android.libraries.ads.mobile.sdk.common.AgeRestrictedTreatment
import com.google.android.libraries.ads.mobile.sdk.common.RequestConfiguration
import com.google.android.libraries.ads.mobile.sdk.initialization.InitializationConfig
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform
import it.amdev.admob.wrapper.diagnostics.IDroidLogger
import it.amdev.admob.wrapper.listeners.OnConsentFormEventListener
import it.amdev.admob.wrapper.listeners.OnConsentGatheringListener
import it.amdev.admob.wrapper.listeners.OnConsentInformationRequestListener
import it.amdev.admob.wrapper.listeners.OnInitializedListener
import it.amdev.admob.wrapper.privacy.AdMobAgeTreatment
import it.amdev.admob.wrapper.privacy.ConsentInformationRequestDebugParameters
import it.amdev.admob.wrapper.privacy.ConsentStatusData
import java.time.Instant

@Suppress("unused")
class AdMobManager(private val logger: IDroidLogger? = null) {

    private var initialized = false
    private var consentInformation: ConsentInformation? = null

    @Volatile
    private var lastConsentRefreshTimestampMilliseconds: Long? = null

    companion object {
        private const val LOG_TAG = "AdMobManager"

        @JvmStatic
        val instance: AdMobManager by lazy { AdMobManager() }
    }

    fun initialize(context: Context,
                   applicationId: String,
                   listener: OnInitializedListener) {
        val initializationConfig = InitializationConfig.Builder(applicationId = applicationId)
                                                       .build()
        initialize(context = context,
                   initializationConfig = initializationConfig,
                   listener = listener)
    }

    fun initialize(context: Context,
                   applicationId: String,
                   ageTreatment: AdMobAgeTreatment,
                   listener: OnInitializedListener) {
        val requestConfiguration = RequestConfiguration.Builder()
                                                       .setAgeRestrictedTreatment(
                                                           ageTreatment.toNativeAgeRestrictedTreatment()
                                                       )
                                                       .build()
        val initializationConfig = InitializationConfig.Builder(applicationId = applicationId)
                                                       .setRequestConfiguration(requestConfiguration)
                                                       .build()

        if (initialized)
            MobileAds.setRequestConfiguration(requestConfiguration)

        initialize(context = context,
                   initializationConfig = initializationConfig,
                   listener = listener)
    }

    fun isInitialized(): Boolean = initialized

    fun updateCurrentConsentInformation(activity: Activity,
                                        tagForUnderAgeOfConsent: Boolean,
                                        listener: OnConsentInformationRequestListener,
                                        requestDebugParameters: ConsentInformationRequestDebugParameters? = null) {
        val consentInformation = this.consentInformation ?: UserMessagingPlatform
                                                            .getConsentInformation(activity.applicationContext)
                                                            .also { this.consentInformation = it }

        var parametersBuilder = ConsentRequestParameters.Builder().setTagForUnderAgeOfConsent(tagForUnderAgeOfConsent)

        requestDebugParameters?.let { debugParameters ->
            val debugSettingsBuilder =
                ConsentDebugSettings.Builder(activity)

            debugParameters.debugGeography?.let {
                debugSettingsBuilder.setDebugGeography(it)
            }

            debugParameters.testDeviceHashedId
                ?.takeIf(String::isNotBlank)
                ?.let(debugSettingsBuilder::addTestDeviceHashedId)

            parametersBuilder = parametersBuilder.setConsentDebugSettings(
                    debugSettingsBuilder.build()
                )
        }

        val parameters = parametersBuilder.build()

        consentInformation.requestConsentInfoUpdate(activity,
                                                parameters,
            {
                lastConsentRefreshTimestampMilliseconds = Instant.now().toEpochMilli()
                listener.onConsentInformationRequestSuccess()
            },
            { error ->
                listener.onConsentInformationRequestFailure(
                    errorCode = error.errorCode,
                    errorMessage = error.message
                )
            }
        )
    }

    fun currentConsentInformation()
        : ConsentStatusData? {
        if (this.consentInformation == null)
            return null

        val currentConsentStatus = this.consentInformation?.consentStatus ?: ConsentInformation.ConsentStatus.UNKNOWN
        val currentPrivacyOptionRequirements = this.consentInformation?.privacyOptionsRequirementStatus ?: ConsentInformation.PrivacyOptionsRequirementStatus.UNKNOWN

        val consentStatusValue: Int = currentConsentStatus

        val privacyOptionRequirementsValue = when(currentPrivacyOptionRequirements) {
            ConsentInformation.PrivacyOptionsRequirementStatus.UNKNOWN -> 0
            ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED -> 1
            ConsentInformation.PrivacyOptionsRequirementStatus.NOT_REQUIRED -> 2
        }

        val epochTimestamp = lastConsentRefreshTimestampMilliseconds ?: 0L
        val consentStatusData = ConsentStatusData(lastRefreshTimestampMilliseconds = epochTimestamp,
                                                  consentStatus = consentStatusValue,
                                                  privacyOptionsRequirementStatus = privacyOptionRequirementsValue)

        return consentStatusData
    }

    fun showPrivacyOptionsForm(activity: Activity,
                               listener: OnConsentFormEventListener)
    {

        val information = consentInformation

        if (information == null) {
            listener.onDismissedWithError(errorCode = -1,
                                          errorMessage =
                                          "Consent information has not been updated.")
            return
        }

        if (information.privacyOptionsRequirementStatus !=
            ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED)
        {
            listener.onDismissedWithError(errorCode = -2,
                                          errorMessage =
                                          "Privacy options form is not required.")
            return
        }

        UserMessagingPlatform.showPrivacyOptionsForm(activity) { formError ->
            if (formError != null) {
                logger?.logError(tag = LOG_TAG,
                                 message = "Error showing privacy options form: ${formError.message}")
                listener.onDismissedWithError(errorCode = formError.errorCode,
                                              errorMessage = formError.message)
            } else {
                logger?.logDebug(tag = LOG_TAG,
                                 message = "Privacy options form dismissed successfully")
                listener.onDismissed()
            }
        }
    }

    fun loadAndShowConsentFormIfRequired(activity: Activity,
                                         listener: OnConsentFormEventListener) {
        UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity)
        { formError ->
            if (formError != null) {
                logger?.logError(tag = LOG_TAG,
                                 message = "Error loading or showing the required consent form: " +
                                           formError.message)
                listener.onDismissedWithError(errorCode = formError.errorCode,
                                              errorMessage = formError.message)
            } else {
                logger?.logDebug(tag = LOG_TAG,
                    message = "Required consent form dismissed successfully")
                listener.onDismissed()
            }
        }
    }

    fun canRequestAds(): Boolean {
        val result = consentInformation?.canRequestAds() ?: false
        return result
    }

    fun gatherConsent(activity: Activity,
                      tagForUnderAgeOfConsent: Boolean,
                      listener: OnConsentGatheringListener,
                      requestDebugParameters: ConsentInformationRequestDebugParameters? = null) {

        updateCurrentConsentInformation(activity = activity,
                                        tagForUnderAgeOfConsent = tagForUnderAgeOfConsent,
                                        requestDebugParameters = requestDebugParameters,
                                        listener = object : OnConsentInformationRequestListener {

                    override fun onConsentInformationRequestSuccess() {
                        UserMessagingPlatform
                            .loadAndShowConsentFormIfRequired(activity) { error ->
                                if (error != null) {
                                    listener.onCompletedWithError(errorCode = error.errorCode,
                                                                  errorMessage = error.message,
                                                                  canRequestAds = consentInformation?.canRequestAds() ?: false,
                                                                  privacyOptionsRequired = consentInformation?.privacyOptionsRequirementStatus == ConsentInformation
                                                                                           .PrivacyOptionsRequirementStatus
                                                                                           .REQUIRED)
                                } else {
                                    listener.onCompleted(canRequestAds = consentInformation?.canRequestAds() ?: false,
                                                         privacyOptionsRequired = consentInformation?.privacyOptionsRequirementStatus == ConsentInformation
                                                                                  .PrivacyOptionsRequirementStatus
                                                                                  .REQUIRED)
                                }
                            }
                    }

                    override fun onConsentInformationRequestFailure(errorCode: Int,
                                                                    errorMessage: String) {
                        listener.onCompletedWithError(errorCode = errorCode,
                                                      errorMessage = errorMessage,
                                                      canRequestAds = consentInformation?.canRequestAds() ?: false,
                                                      privacyOptionsRequired = consentInformation?.privacyOptionsRequirementStatus == ConsentInformation
                                                                               .PrivacyOptionsRequirementStatus
                                                                               .REQUIRED)
                    }
                }
        )
    }

    fun resetConsentForTesting() {
        consentInformation?.reset()
        lastConsentRefreshTimestampMilliseconds = null
    }

    private fun initialize(context: Context,
                           initializationConfig: InitializationConfig,
                           listener: OnInitializedListener) {
        if (initialized) {
            listener.onInitialized()
            return
        }
        Thread {
            try {
                MobileAds.initialize(context, initializationConfig) {
                    initialized = true
                    listener.onInitialized()
                }
            } catch (e: Exception) {
                listener.onInitializationFailed(e.message ?: "Unknown error")
            }
        }.start()
    }

    private fun AdMobAgeTreatment.toNativeAgeRestrictedTreatment(): AgeRestrictedTreatment =
        when (this) {
            AdMobAgeTreatment.Unspecified -> AgeRestrictedTreatment.UNSPECIFIED
            AdMobAgeTreatment.Child -> AgeRestrictedTreatment.CHILD
            AdMobAgeTreatment.Teen -> AgeRestrictedTreatment.TEEN
        }
}

File: sources/apple/ios/AdMobWrapper/AdMobManager.swift
//
//  AdMobManager.swift
//  AdMobWrapper
//
//  Created by Alessandro Morvillo on 20/04/26.
//

import Foundation
import GoogleMobileAds
import UIKit
import UserMessagingPlatform

@objc public class AdMobManager: NSObject {

    private static let logTag = "AdMobManager"

    @objc public static let instance = AdMobManager()

    private var initialized = false
    private var consentInformationAvailable = false
    private var lastConsentRefreshTimestampMilliseconds: Int64?
    private var logger: IAppleLogger?

    private override init() {
        self.logger = nil
        super.init()
    }

    @objc public convenience init(logger: IAppleLogger?) {
        self.init()
        self.logger = logger
    }

    @objc public func initialize(viewController: UIViewController,
                                 listener: OnInitializedListener) {
        self.initializeInternal(viewController: viewController,
                                ageTreatment: nil,
                                listener: listener)
    }

    @objc public func initialize(viewController: UIViewController,
                                 ageTreatment: AdMobAgeTreatment,
                                 listener: OnInitializedListener) {
        self.initializeInternal(viewController: viewController,
                                ageTreatment: ageTreatment,
                                listener: listener)
    }

    @objc public func isInitialized() -> Bool {
        return self.executeOnMainAndWait {
            self.initialized
        }
    }

    @objc public func updateCurrentConsentInformation(
        viewController: UIViewController,
        tagForUnderAgeOfConsent: Bool,
        listener: OnConsentInformationRequestListener
    ) {
        self.updateCurrentConsentInformation(
            viewController: viewController,
            tagForUnderAgeOfConsent: tagForUnderAgeOfConsent,
            listener: listener,
            requestDebugParameters: nil
        )
    }

    @objc public func updateCurrentConsentInformation(
        viewController: UIViewController,
        tagForUnderAgeOfConsent: Bool,
        listener: OnConsentInformationRequestListener,
        requestDebugParameters: ConsentInformationRequestDebugParameters?
    ) {
        self.executeOnMain { [weak self] in
            guard let self = self else { return }

            self.consentInformationAvailable = true

            let parameters = self.createConsentRequestParameters(
                tagForUnderAgeOfConsent: tagForUnderAgeOfConsent,
                requestDebugParameters: requestDebugParameters
            )

            ConsentInformation.shared.requestConsentInfoUpdate(with: parameters) {
                [weak self] requestConsentError in
                guard let self = self else { return }

                if let requestConsentError = requestConsentError {
                    let error = requestConsentError as NSError
                    listener.onConsentInformationRequestFailure(
                        errorCode: error.code,
                        errorMessage: error.localizedDescription
                    )
                    return
                }

                self.lastConsentRefreshTimestampMilliseconds = Int64(
                    Date().timeIntervalSince1970 * 1_000
                )
                listener.onConsentInformationRequestSuccess()
            }
        }
    }

    @objc public func currentConsentInformation() -> ConsentStatusData? {
        return self.executeOnMainAndWait {
            guard self.consentInformationAvailable else { return nil }

            let consentInformation = ConsentInformation.shared
            return ConsentStatusData(
                lastRefreshTimestampMilliseconds: self.lastConsentRefreshTimestampMilliseconds ?? 0,
                consentStatus: consentInformation.consentStatus.rawValue,
                privacyOptionsRequirementStatus: consentInformation.privacyOptionsRequirementStatus.rawValue
            )
        }
    }

    @objc public func showPrivacyOptionsForm(viewController: UIViewController,
                                             listener: OnConsentFormEventListener) {
        Task { @MainActor [weak self] in
            guard let self = self else { return }

            guard self.consentInformationAvailable else {
                listener.onDismissedWithError(
                    errorCode: -1,
                    errorMessage: "Consent information has not been updated."
                )
                return
            }

            guard ConsentInformation.shared.privacyOptionsRequirementStatus == .required else {
                listener.onDismissedWithError(
                    errorCode: -2,
                    errorMessage: "Privacy options form is not required."
                )
                return
            }

            do {
                try await ConsentForm.presentPrivacyOptionsForm(from: viewController)
                self.logger?.logDebug(
                    message: "Privacy options form dismissed successfully",
                    tag: Self.logTag
                )
                listener.onDismissed()
            } catch {
                let error = error as NSError
                self.logger?.logError(
                    message: "Error showing privacy options form: \(error.localizedDescription)",
                    tag: Self.logTag
                )
                listener.onDismissedWithError(
                    errorCode: error.code,
                    errorMessage: error.localizedDescription
                )
            }
        }
    }

    @objc public func loadAndShowConsentFormIfRequired(
        viewController: UIViewController,
        listener: OnConsentFormEventListener
    ) {
        Task { @MainActor [weak self] in
            guard let self = self else { return }

            do {
                try await ConsentForm.loadAndPresentIfRequired(from: viewController)
                self.logger?.logDebug(
                    message: "Required consent form dismissed successfully",
                    tag: Self.logTag
                )
                listener.onDismissed()
            } catch {
                let error = error as NSError
                self.logger?.logError(
                    message: "Error loading or showing the required consent form: " +
                             error.localizedDescription,
                    tag: Self.logTag
                )
                listener.onDismissedWithError(
                    errorCode: error.code,
                    errorMessage: error.localizedDescription
                )
            }
        }
    }

    @objc public func canRequestAds() -> Bool {
        return self.executeOnMainAndWait {
            guard self.consentInformationAvailable else { return false }
            return ConsentInformation.shared.canRequestAds
        }
    }

    @objc public func gatherConsent(
        viewController: UIViewController,
        tagForUnderAgeOfConsent: Bool,
        listener: OnConsentGatheringListener
    ) {
        self.gatherConsent(
            viewController: viewController,
            tagForUnderAgeOfConsent: tagForUnderAgeOfConsent,
            listener: listener,
            requestDebugParameters: nil
        )
    }

    @objc public func gatherConsent(
        viewController: UIViewController,
        tagForUnderAgeOfConsent: Bool,
        listener: OnConsentGatheringListener,
        requestDebugParameters: ConsentInformationRequestDebugParameters?
    ) {
        let requestListener = ConsentInformationRequestListenerAdapter(
            onSuccess: { [weak self] in
                guard let self = self else { return }

                self.loadAndShowConsentFormIfRequired(
                    viewController: viewController,
                    listener: ConsentFormEventListenerAdapter(
                        onDismissed: {
                            listener.onCompleted(
                                canRequestAds: self.canRequestAds(),
                                privacyOptionsRequired: self.isPrivacyOptionsRequired()
                            )
                        },
                        onDismissedWithError: { errorCode, errorMessage in
                            listener.onCompletedWithError(
                                errorCode: errorCode,
                                errorMessage: errorMessage ?? "Unknown error",
                                canRequestAds: self.canRequestAds(),
                                privacyOptionsRequired: self.isPrivacyOptionsRequired()
                            )
                        }
                    )
                )
            },
            onFailure: { [weak self] errorCode, errorMessage in
                guard let self = self else { return }

                listener.onCompletedWithError(
                    errorCode: errorCode,
                    errorMessage: errorMessage,
                    canRequestAds: self.canRequestAds(),
                    privacyOptionsRequired: self.isPrivacyOptionsRequired()
                )
            }
        )

        self.updateCurrentConsentInformation(
            viewController: viewController,
            tagForUnderAgeOfConsent: tagForUnderAgeOfConsent,
            listener: requestListener,
            requestDebugParameters: requestDebugParameters
        )
    }

    @objc public func resetConsentForTesting() {
        self.executeOnMainAndWait {
            ConsentInformation.shared.reset()
            self.lastConsentRefreshTimestampMilliseconds = nil
        }
    }

    private func initializeInternal(viewController: UIViewController,
                                    ageTreatment: AdMobAgeTreatment?,
                                    listener: OnInitializedListener) {
        self.executeOnMain { [weak self] in
            guard let self = self else { return }

            if let ageTreatment = ageTreatment {
                MobileAds.shared.requestConfiguration.ageRestrictedTreatment =
                    ageTreatment.nativeAgeRestrictedTreatment
            }

            if self.initialized {
                listener.onInitialized()
                return
            }

            MobileAds.shared.start { [weak self] _ in
                guard let self = self else { return }
                self.initialized = true
                listener.onInitialized()
            }
        }
    }

    private func createConsentRequestParameters(
        tagForUnderAgeOfConsent: Bool,
        requestDebugParameters: ConsentInformationRequestDebugParameters?
    ) -> RequestParameters {
        let parameters = RequestParameters()
        parameters.isTaggedForUnderAgeOfConsent = tagForUnderAgeOfConsent

        if let requestDebugParameters = requestDebugParameters {
            let debugSettings = DebugSettings()

            if let geographyValue = requestDebugParameters.debugGeography?.intValue,
               let geography = DebugGeography(rawValue: geographyValue) {
                debugSettings.geography = geography
            }

            if let testDeviceHashedId = requestDebugParameters.testDeviceHashedId,
               !testDeviceHashedId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                debugSettings.testDeviceIdentifiers = [testDeviceHashedId]
            }

            parameters.debugSettings = debugSettings
        }

        return parameters
    }

    private func isPrivacyOptionsRequired() -> Bool {
        return self.executeOnMainAndWait {
            guard self.consentInformationAvailable else { return false }
            return ConsentInformation.shared.privacyOptionsRequirementStatus == .required
        }
    }

    private func executeOnMain(_ action: @escaping () -> Void) {
        if Thread.isMainThread {
            action()
        } else {
            DispatchQueue.main.async(execute: action)
        }
    }

    @discardableResult
    private func executeOnMainAndWait<T>(_ action: () -> T) -> T {
        if Thread.isMainThread {
            return action()
        }

        return DispatchQueue.main.sync(execute: action)
    }
}

private extension AdMobAgeTreatment {
    var nativeAgeRestrictedTreatment: AgeRestrictedTreatment {
        switch self {
        case .unspecified:
            return .unspecified
        case .child:
            return .child
        case .teen:
            return .teen
        }
    }
}

private final class ConsentInformationRequestListenerAdapter: NSObject,
                                                              OnConsentInformationRequestListener {
    private let onSuccess: () -> Void
    private let onFailure: (Int, String) -> Void

    init(onSuccess: @escaping () -> Void,
         onFailure: @escaping (Int, String) -> Void) {
        self.onSuccess = onSuccess
        self.onFailure = onFailure
        super.init()
    }

    func onConsentInformationRequestSuccess() {
        self.onSuccess()
    }

    func onConsentInformationRequestFailure(errorCode: Int,
                                             errorMessage: String) {
        self.onFailure(errorCode, errorMessage)
    }
}

private final class ConsentFormEventListenerAdapter: NSObject,
                                                     OnConsentFormEventListener {
    private let onDismissedHandler: () -> Void
    private let onDismissedWithErrorHandler: (Int, String?) -> Void

    init(onDismissed: @escaping () -> Void,
         onDismissedWithError: @escaping (Int, String?) -> Void) {
        self.onDismissedHandler = onDismissed
        self.onDismissedWithErrorHandler = onDismissedWithError
        super.init()
    }

    func onDismissed() {
        self.onDismissedHandler()
    }

    func onDismissedWithError(errorCode: Int,
                              errorMessage: String?) {
        self.onDismissedWithErrorHandler(errorCode, errorMessage)
    }
}
